drivers/display/ssd16xx.c

```diff
@@ -974,6 +975,44 @@ static int ssd16xx_init(const struct device *dev)
 	return ssd16xx_controller_init(dev);
 }
 
+#ifdef CONFIG_PM_DEVICE
+static int ssd16xx_pm_action(const struct device *dev,
+			     enum pm_device_action action)
+{
+	int err = 0;
+	const struct ssd16xx_config *config = dev->config;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		err = ssd16xx_controller_init(dev);
+		if (err < 0) {
+			return err;
+		}
+		break;
+
+	case PM_DEVICE_ACTION_SUSPEND:
+		ssd16xx_busy_wait(dev);
+		err = ssd16xx_write_uint8(dev, SSD16XX_CMD_SLEEP_MODE, 1);
+		if (err < 0) {
+			return err;
+		}
+
+		k_msleep(SSD16XX_BUSY_DELAY);
+
+		err = gpio_pin_set_dt(&config->reset_gpio, 1);
```

Comment:
Is this necessary? I have been able to get good results without holding the device in reset. It's probably even undesirable if you are dropping into SRAM retention in which case you can resume and perform a partial refresh.

Reply:
It might on some boards, on the `Waveshare 296x128, 2.9inch E-Ink display module` I tested on, it's definitely required to assert the reset. Only with the reset asserted, I get the negligible consumption as measured here: https://github.com/fgervais/project-nrf-dishwasher-sign?tab=readme-ov-file#power-consumption

Reply:
When developing the patch, I measured power consumption in sleep and got the expected results. I don't know for sure if that was on a SSD1680 though. It's worth pointing out that the data sheet doesn't mention anything about holding the chip in reset to enter deep sleep which is another reason I find it a bit suspicious. I have a vague recollection of something being somewhat quirky though which might be why I had to activate the clock manually.

Could you give my original patch a go and see if it works for you? I'd much prefer that solution since it works as expected with partial refresh.

Reply:
Yes I will give it a try. I don't have the test setup available anymore so I will have to put something together. I will report here my results.

Reply:
I did the test with this code: https://github.com/andysan/zephyr/blob/430094e6d78ec72dab06fd7ea620735d865c8e44/drivers/display/ssd16xx.c

And I get just below 2.5mA average consumption.

In this screenshot, It's suspended, then I trigger a full refresh and it suspends again. (Ref. code: https://github.com/fgervais/project-nrf-dishwasher-sign/blob/99a7a846b6771a6f46df6ceafe36de45a9151a18/app/src/main.c#L134)

![Screenshot from 2024-06-09 20-59-04](https://github.com/zephyrproject-rtos/zephyr/assets/5411996/006f9f32-714e-4aaf-94c8-d88cf7e66729)

Reply:
I did more tests and the power consumption seems to be coming from the adapter board and not the display itself.

Pulling the RST signal low is basically just working around a mistake with the specific Waveshare display module I'm using.

I will proceed with a little hardware fix on the module and I should then be able to properly test the suggested suspend implementation.

Reply:
Thanks for looking into this. I can't remember which module I used when developing my old patch, but that particular module didn't seem to exhibit this behaviour. It might be worth having the option of keeping the module in reset, but that seems like it should be a quirk that is either enabled using a specific compat string or a device tree option.

Reply:
Well I did reduce consumption by 10 but I'm still at 246uA. Basically this module is not made with low power consumption in mind.

With this kind of consumption I cannot tell if the display is sleeping as it should or not and so I think won't be able to perform the test with what I have on hand right now.